council/llm/llm_function/llm_pipeline.py

```diff
@@ -0,0 +1,219 @@
+from __future__ import annotations
+
+from typing import Any, Dict, Final, Generic, List, Optional, Protocol, Sequence, Type, TypeVar, Union, cast
+
+from council.llm.base import LLMBase, LLMMessage
+
+from .llm_function import LLMFunction
+from .llm_middleware import LLMMiddlewareChain
+from .llm_response_parser import BaseModelResponseParser
+
+
+class ProcessorException(Exception):
+    """
+    Exception raised during the execution of a Processor.
+    Contains information about the input that caused the exception, the exception itself,
+    and optionally name of a previous processor that should handle the exception.
+    """
+
+    def __init__(self, *, input: str, message: str, transfer_to: Optional[str] = None):
+        self.input = input
+        self.message = message
+        self.transfer_to = transfer_to
+
+
+class LLMProcessorInput(Protocol):
+    """Input for a LLMProcessor."""
+
+    def to_prompt(self) -> str:
+        """Convert the object to a prompt string."""
+        ...
+
+
+T_Input = TypeVar("T_Input")
+T_Output = TypeVar("T_Output")
+
+# input must implement to_prompt()
+T_LLMInput = TypeVar("T_LLMInput", bound=LLMProcessorInput)
+# output must implement from_response() and to_response_template()
+T_LLMOutput = TypeVar("T_LLMOutput", bound=BaseModelResponseParser)
+
+
+class LLMProcessorRecord:
+    """Record of a LLMProcessor execution."""
+
+    def __init__(self, *, input: str, output: str, exception: Optional[ProcessorException] = None):
+        self.input = input
+        self.output = output
+        self.exception = exception
+
+    def set_exception(self, exception: ProcessorException) -> None:
+        self.exception = exception
+
+    def to_dict(self) -> Dict[str, str]:
+        result = {
+            "input": self.input,
+            "output": self.output,
+        }
+        if self.exception is not None:
+            result["exception"] = self.exception.message
+        return result
+
+
+class Processor(Generic[T_Input, T_Output]):
```

Comment:
Could be a base class:
```suggestion
class ProcessorBase(Generic[T_Input, T_Output], abc.ABC):
```